### DISCUSSION TITLE
Controller manager not available

### DISCUSSION THREAD

**Discussion by jarain78** [iRobotEducation/create3_docs#394] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Ubuntu 22.04

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.1.2

### Which RMW is your robot running?

CycloneDDS

### Is the robot connected to a network? If so, what is the network type?

yes

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

Hello everyone,

I'm currently attempting to launch the Gazebo simulation of Create 3 by using the following command: _ros2 launch irobot_create_gazebo_bringup create3_gazebo_aws_small.launch.py_. Gazebo and rviz2 start up successfully, but I'm encountering some errors in the command terminal:

> [spawn_entity.py-8] [INFO] [1685264678.932643621] [spawn_create3]: Calling service /spawn_entity
> [spawn_entity.py-5] [INFO] [1685264678.944124495] [spawn_standard_dock]: Calling service /spawn_entity
> [gzserver-1] [INFO] [1685264679.145328122] [cliff_front_left]: Started plugin
> [gzserver-1] [INFO] [1685264679.150227469] [cliff_front_right]: Started plugin
> [gzserver-1] [INFO] [1685264679.155056887] [cliff_side_left]: Started plugin
> [gzserver-1] [INFO] [1685264679.160126254] [cliff_side_right]: Started plugin
> [gzserver-1] [INFO] [1685264679.171753475] [bumper_plugin]: Bumper plugin loaded correctly
> [gzserver-1] [INFO] [1685264679.194560221] [ir_intensity_front_center_left]: Starting IR Emitter Plugin!
> [gzserver-1] [INFO] [1685264679.199682857] [ir_intensity_front_center_right]: Starting IR Emitter Plugin!
> [gzserver-1] [INFO] [1685264679.206956352] [ir_intensity_front_left]: Starting IR Emitter Plugin!
> [gzserver-1] [INFO] [1685264679.214698477] [ir_intensity_front_right]: Starting IR Emitter Plugin!
> [gzserver-1] [INFO] [1685264679.223139032] [ir_intensity_left]: Starting IR Emitter Plugin!
> [gzserver-1] [INFO] [1685264679.231484046] [ir_intensity_right]: Starting IR Emitter Plugin!
> [spawn_entity.py-8] [INFO] [1685264679.239775211] [spawn_create3]: Spawn status: SpawnEntity: Successfully spawned entity [create3]
> [gzserver-1] [INFO] [1685264679.240458740] [ir_intensity_side_left]: Starting IR Emitter Plugin!
> [gzserver-1] [INFO] [1685264679.262690805] [wheel_drop_left_plugin]: Started wheel drop plugin
> [gzserver-1] [INFO] [1685264679.267237543] [wheel_drop_right_plugin]: Started wheel drop plugin
> [gzserver-1] [INFO] [1685264679.278959296] [mouse_plugin]: Starting optical mouse plugin
> [gzserver-1] [INFO] [1685264679.290352637] [ir_omni_plugin]: Starting ir opcode plugin
> [gzserver-1] [INFO] [1685264679.315341322] [dock_status_publisher]: Starting ir opcode plugin
> [gzserver-1] [WARN] [1685264679.316079872] [dock_status_publisher]: standard_dock model is not ready yet
> [INFO] [spawn_entity.py-8]: process has finished cleanly [pid 92198]
> [spawn_entity.py-5] [INFO] [1685264679.589437663] [spawn_standard_dock]: Spawn status: SpawnEntity: Successfully spawned entity [standard_dock]
> [INFO] [spawn_entity.py-5]: process has finished cleanly [pid 92192]
> [spawner-9] [INFO] [1685264680.605728497] [spawner_joint_state_broadcaster]: Waiting for '/controller_manager' node to exist
> [spawner-9] [INFO] [1685264682.612025255] [spawner_joint_state_broadcaster]: Waiting for '/controller_manager' node to exist
> [spawner-9] [INFO] [1685264684.618439824] [spawner_joint_state_broadcaster]: Waiting for '/controller_manager' node to exist
> [spawner-9] [ERROR] [1685264686.624387024] [spawner_joint_state_broadcaster]: Controller manager not available
> [ERROR] [spawner-9]: process has died [pid 92200, exit code 1, cmd '/home/jarain78/ros2_ws/install/controller_manager/lib/controller_manager/spawner joint_state_broadcaster -c controller_manager --ros-args -r __ns:=/'].
> [INFO] [spawner-19]: process started with pid [92606]
> [spawner-19] [INFO] [1685264689.090404302] [spawner_diffdrive_controller]: Waiting for '/controller_manager' node to exist
> [spawner-19] [INFO] [1685264691.096688792] [spawner_diffdrive_controller]: Waiting for '/controller_manager' node to exist
> [spawner-19] [INFO] [1685264693.102573452] [spawner_diffdrive_controller]: Waiting for '/controller_manager' node to exist
> [spawner-19] [INFO] [1685264695.108660962] [spawner_diffdrive_controller]: Waiting for '/controller_manager' node to exist
> [spawner-19] [ERROR] [1685264697.114772843] [spawner_diffdrive_controller]: Controller manager not available
> [ERROR] [spawner-19]: process has died [pid 92606, exit code 1, cmd '/home/jarain78/ros2_ws/install/controller_manager/lib/controller_manager/spawner diffdrive_controller -c controller_manager --ros-args -r __ns:=/ --params-file /home/jarain78/ros2_ws/install/irobot_create_control/share/irobot_create_control/config/control.yaml'].
> 

I suspect that these errors are causing the robot to not respond to the \dock and \undock actions. When I try to execute these actions, the system responds with: "Undock Goal Aborted."

I would greatly appreciate it if someone could provide assistance in resolving this issue.

Thank you very much.

Best regards,

**Comment by justinIRBT** [comment]
Controller manager comes from the external dependency `ros2_control`.  Can you confirm it is installed on your system?  It should have been pulled in during the 
```
rosdep install --from-path src -yi
```
step.  You are correct that without the controller manager, the robot won't run correctly.

**Comment by justinIRBT** [comment]
Can you attach the full text of the application startup from when you initiate the command?  I wonder if something is dying earlier in the startup process, sometimes it can be hidden because gazebo is starting up so many things simultaneously.  I just tested a fresh checkout of galactic on my computer and it started up without issues.

**Reply by justinIRBT** [reply]
If the issue is CPU load and things not starting up on time, can you try running this diff (I applied this for galactic, so you will need to see if the arguments are the same on humble)
```
diff --git a/irobot_create_common/irobot_create_control/launch/include/control.py b/irobot_create_common/irobot_create_control/launch/include/control.py
index f8a781d..cca6994 100644
--- a/irobot_create_common/irobot_create_control/launch/include/control.py
+++ b/irobot_create_common/irobot_create_control/launch/include/control.py
@@ -21,14 +21,14 @@ def generate_launch_description():
         package='controller_manager',
         executable='spawner',
         parameters=[control_params_file],
-        arguments=['diffdrive_controller', '-c', '/controller_manager'],
+        arguments=['diffdrive_controller', '-c', '/controller_manager', '--controller-manager-timeout', str(35)],
         output='screen',
     )
 
     joint_state_broadcaster_spawner = Node(
         package='controller_manager',
         executable='spawner',
-        arguments=['joint_state_broadcaster', '-c', '/controller_manager'],
+        arguments=['joint_state_broadcaster', '-c', '/controller_manager', '--controller-manager-timeout', str(35)],
         output='screen',
     )
 
diff --git a/irobot_create_gazebo/irobot_create_gazebo_bringup/launch/create3_gazebo.launch.py b/irobot_create_gazebo/irobot_create_gazebo_bringup/launch/create3_gazebo.launch.py
index fd9a532..87bbddf 100644
--- a/irobot_create_gazebo/irobot_create_gazebo_bringup/launch/create3_gazebo.launch.py
+++ b/irobot_create_gazebo/irobot_create_gazebo_bringup/launch/create3_gazebo.launch.py
@@ -92,7 +92,7 @@ def generate_launch_description():
     x, y, z = LaunchConfiguration('x'), LaunchConfiguration('y'), LaunchConfiguration('z')
     yaw = LaunchConfiguration('yaw')
     world_path = LaunchConfiguration('world_path')
-    spawn_dock = LaunchConfiguration('spawn_dock')
+    should_spawn_dock = LaunchConfiguration('spawn_dock')
     use_gazebo_gui = LaunchConfiguration('use_gazebo_gui')
     use_rviz = LaunchConfiguration('use_rviz')
 
@@ -120,7 +120,7 @@ def generate_launch_description():
     yaw_dock = OffsetParser(yaw, 3.1416)
     dock_description = IncludeLaunchDescription(
         PythonLaunchDescriptionSource([dock_description_launch_file]),
-        condition=IfCondition(spawn_dock),
+        condition=IfCondition(should_spawn_dock),
         # The robot starts docked
         launch_arguments={'x': x_dock, 'y': y, 'z': z, 'yaw': yaw_dock}.items(),
     )
@@ -135,9 +135,10 @@ def generate_launch_description():
                    '-x', x_dock,
                    '-y', y,
                    '-z', z,
-                   '-Y', yaw_dock],
+                   '-Y', yaw_dock,
+                   '-spawn_service_timeout', str(35)],
         output='screen',
-        condition=IfCondition(spawn_dock),
+        condition=IfCondition(should_spawn_dock),
     )
 
     # Create 3 robot model and description
@@ -155,7 +156,8 @@ def generate_launch_description():
                    '-x', x,
                    '-y', y,
                    '-z', z,
-                   '-Y', yaw],
+                   '-Y', yaw,
+                   '-spawn_service_timeout', str(35)],
         output='screen',
     )
```
